### ISSUE TITLE
Context lost between HttpClientWrapper and DelegatingHandler as of 15.0.7

### ISSUE THREAD

**Issue by amweiss** [ThreeMammals/Ocelot#1252]
## Expected Behavior / New Feature
A DelegatingHandler with `IHttpContextAccessor` injected can access the current context correctly.

## Actual Behavior / Motivation for New Feature
As of version 15.0.7, a custom `DelegatingHandler` we use now only has access to an empty context and thus `httpContextAccessor.HttpContext.User.FindFirst(ClaimTypes.NameIdentifier)?.Value` is now null instead of the name like it was in 15.0.6

## Steps to Reproduce the Problem

  1. Create a `DelegatingHandler` that takes in `IHttpContextAccessor` as an injected dependency. 
  1. Register the handler on the route.
  1. Add JWT auth to the route.
  1. In the `SendAsync` method, call `httpContextAccessor.HttpContext.User.FindFirst(ClaimTypes.NameIdentifier)?.Value`
  1. See that the value is null

## Specifications

  - Version: 15.0.7
  - Platform: dotnet 3.1.300
  - Subsystem: MacOS

**Comment by davipsr** [comment]
Same issue here, version 16.0.1

**Comment by jlukawska** [comment]
Hello, 
The reason is that after changes done in fe3e8bd23a666daa31ed2e8b00dce2c4630c23a7 a new HttpContext instance is created for every request (or even more instances for aggregated routes). AuthenticationMiddleware (and others) receives a new HttpContext instance, but HttpContextAccessor still has a reference to the original instance. That's why HttpContext obtained by HttpContextAccessor doesn't have any claims.

I've proposed a solution (see pull request #1268 ) where the proper HttpContext will be added as a property to a delegating handler class if the class implements the specified interface.

**Comment by bjartekh** [comment]
I tried the code submitted in #1268 (by implementing IDelegatingHandlerWithHttpContext in my DelegatingHandler)  but I experienced that HttpContext.User.Claims were not updated after the first request (so any subsequent requests would use the same user claims). 

Steps to reproduce:
* Obtain an accessToken for ClientID A (from IdentityServer4)
* Use accessToken A towards Route 1 in Ocelot
* Observe that HttpContext.User.Claims contained claims present in accessToken A

* Obtain an accessToken for ClientID B (from IdentityServer4)
* Use accessToken B towards Route 1 in Ocelot
* Observe that HttpContext.User.Claims contained claims present in accessToken A

Is this a scenario you've tested? 

My workaround-hack for getting user claims to be available in the `DelegatingHandler`, was to fetch the claims from the httpContext in `PreAuthorizationMiddleware` and then set them in a custom request scoped RequestData object. 

And when I needed that data in the `DelegatingHandler`, I would just obtain it like this:
`var rqData = (RequestData)_httpContextAccessor.HttpContext?.RequestServices.GetService(typeof(RequestData));`

**Comment by jlukawska** [comment]
@bjartekh thanks, you're right. There's a problem with caching. I'll look into it if I find time.

Unfortunately Ocelot seems to be abandoned after breaking changes with HttpContext :/

**Comment by jlukawska** [comment]
@bjartekh if you are interested, I've just fixed the problem with cache.

**Comment by AgentShark** [comment]
The Ocelot project seems dead, but if anyone experiences this issue, here is a workaround. 

**Set and use a `IHttpContextAccessor`.**

1. Register the `IHttpContextAccessor` dependency.
```CS
// Startup.cs
services.AddHttpContextAccessor();
```
2. In your Ocelot configuration `OcelotPipelineConfiguration` add this code to any Ocelot pipeline middleware that runs **after** `AuthenticationMiddleware`. Example used is `PreQueryStringBuilderMiddleware`, but you can also use `PreAuthorizationMiddleware` or `AuthorizationMiddleware`.
```CS
// Startup.cs
app.UseOcelot(om => {
    om.PreQueryStringBuilderMiddleware = async (context, next) => {
        var contextAccessor = context.RequestServices.GetRequiredService<IHttpContextAccessor>();
        contextAccessor.HttpContext = context;
    }
}).Wait();
```

3. Then inject IHttpContextAccessor into any `DelegatingHandler` as normal.

**Comment by raman-m** [comment]
Hi @AgentShark !
Thanks for the nice coding recipe!

Does it solve the problem with user claims (aka `User` property)?

What about to make it as an additional method for the [IOcelotBuilder](../blob/develop/src/Ocelot/DependencyInjection/IOcelotBuilder.cs) interface of [DependencyInjection](../tree/develop/src/Ocelot/DependencyInjection) part of Ocelot?

**Comment by AgentShark** [comment]
@raman-m Yes it solves the problems with user claims. `AddHttpContextAccessor()` can be registered anywhere in services like any other dependency. I don't think there is any such thing as `UseHttpContextAccessor()`. See [Docs](https://learn.microsoft.com/en-us/aspnet/core/fundamentals/http-context?view=aspnetcore-7.0#access-httpcontext-from-custom-components)

**Comment by raman-m** [comment]
@amweiss  Hi Adam!

I would like personally to thank you for reporting this issue being found in v15.x! Yeah, it was a large upgrade to .NET 5 with a lot of breaking changes made by Tom.

Are you still interested to collaborate on this issue?

>## Specifications
>  - **Version: 15.0.7**
>  - Platform: dotnet 3.1.300
>  - Subsystem: MacOS

Are you able to upgrade your solution to .NET 7 with usage of Ocelot v19.0.2 (latest ver.)?
Could you report that the bug is still in v19.0.2 please?

@davipsr  and @bjartekh  
Could I pleasantly ask you to check the latest v19.0.2 please?

@AgentShark 
What version of Ocelot do you use for your solution proposed in [comment on June 26, 2023](#issuecomment-1606496115)?

**Comment by amweiss** [comment]
I no longer use this project so I won't be going back to verify.

**Comment by raman-m** [comment]
@amweiss 
This is last notification for you...
> I no longer use this project so I won't be going back to verify.

Sorry for disturbing you! You can unsubscribe from notifications for this issue.
And I won't address you in discussions anymore.

**Comment by AgentShark** [comment]
@raman-m I use v18.0.0 where the issue is still present.

**Comment by raman-m** [comment]
@AgentShark wrote [on June 26, 2023](#issuecomment-1608208592):
> Yes it solves the problem with user claims.

How can we verify your solution? Do you have some time for contribution?
Could you upload your app sample to GitHub please, as a separate repo?
Also, you can fork the Ocelot repo, create feature branch, and upload your solution as sample in [samples](https://github.com/ThreeMammals/Ocelot/tree/develop/samples) folder.
And, please, make a reference to Ocelot project to be sure we test the latest version (.NET 7, v19.0.2).

---
> I don't think there is any such thing as `UseHttpContextAccessor()`. See [Docs](https://learn.microsoft.com/en-us/aspnet/core/fundamentals/http-context?view=aspnetcore-7.0#access-httpcontext-from-custom-components)

Okay... It seems you misunderstood me.
I meant that we can wrap this your code:
```csharp
app.UseOcelot(oc =>
{
    oc.PreQueryStringBuilderMiddleware = async (context, next) => 
    {
        var contextAccessor = context.RequestServices.GetRequiredService<IHttpContextAccessor>();
        contextAccessor.HttpContext = context;
        await next.Invoke();
    };
}).Wait();
```
...as `UseAAA()` extension for the `IApplicationBuilder` interface in the [OcelotMiddlewareExtensions](../blob/develop/src/Ocelot/Middleware/OcelotMiddlewareExtensions.cs) class.
And/Or
We can develop `AddAAA()` method for the [OcelotBuilder](../blob/develop/src/Ocelot/DependencyInjection/OcelotBuilder.cs) class, which will wrap `services.AddHttpContextAccessor()` and Ocelot Middleware Configuration as in the [Delegating Handlers](https://ocelot.readthedocs.io/en/latest/features/delegatinghandlers.html?highlight=AddDelegatingHandler) feature.
I believe we have to extend the [OcelotPipelineConfiguration](https://github.com/ThreeMammals/Ocelot/blob/develop/src/Ocelot/Middleware/OcelotPipelineConfiguration.cs#L10) class.

Finally we need to update Ocelot docs:
- [Middleware Injection and Overrides](https://ocelot.readthedocs.io/en/latest/features/middlewareinjection.html)
- [Delegating Handlers](https://ocelot.readthedocs.io/en/latest/features/delegatinghandlers.html)

---
> See [Docs](https://learn.microsoft.com/en-us/aspnet/core/fundamentals/http-context?view=aspnetcore-7.0#access-httpcontext-from-custom-components) (Access `HttpContext` from custom components)

Microsoft Learn: [Access HttpContext in ASP.NET Core](https://learn.microsoft.com/en-us/aspnet/core/fundamentals/http-context?view=aspnetcore-7.0)
Exactly! I would say we need to follow the best practices by Microsoft.
I believe your solution is easier one and solid, but it requires some project update with your proposed functionality.

**Comment by raman-m** [comment]
TODO
Create new issue as feature request to design a middleware attached to Ocelot pipeline as the last step before request sending.

**Comment by raman-m** [comment]
```diff
+ Accepted 
```
...because of ready PR #1268

**Comment by raman-m** [comment]
@ggnaegi, FYI, the linked pull request #1268 was closed a few days ago due to a significant number of merge conflicts.
I require your assistance to determine the appropriate action for this issue. Since we've implemented the new `MessageInvokerPool` routing design in the system kernel, it appears this issue may no longer be pertinent. Ultimately, we must replicate the bug testing on the most recent version.

**Comment by ggnaegi** [comment]
@raman-m Do we have this issue in a multiplexing context?
